contracts/contracts/strategies/uniswap/UniswapV3LiquidityManager.sol

```diff
@@ -0,0 +1,830 @@
+        if (token0Balance < desiredAmount0) {
+            vault.withdrawFromUniswapV3Reserve(
+                token0,
+                desiredAmount0 - token0Balance
+            );
+        }
+
+        uint256 token1Balance = IERC20(token1).balanceOf(address(this));
+        if (token1Balance < desiredAmount1) {
+            vault.withdrawFromUniswapV3Reserve(
+                token1,
+                desiredAmount1 - token1Balance
+            );
+        }
+
+        // TODO: Check value of assets moved here
+    }
+
+    function _checkSwapLimits(uint160 sqrtPriceLimitX96, bool swapZeroForOne)
+        internal
+    {
+        require(!swapsPaused, "Swaps are paused");
+
+        (uint160 currentPriceX96, , , , , , ) = pool.slot0();
+
+        if (minSwapPriceX96 > 0 || maxSwapPriceX96 > 0) {
+            require(
+                minSwapPriceX96 <= currentPriceX96 &&
+                    currentPriceX96 <= maxSwapPriceX96,
+                "Price out of bounds"
+            );
+
+            require(
+                swapZeroForOne
+                    ? (sqrtPriceLimitX96 >= minSwapPriceX96)
+                    : (sqrtPriceLimitX96 <= maxSwapPriceX96),
+                "Slippage out of bounds"
+            );
+        }
+    }
+
+    function _ensureAssetsBySwapping(
+        uint256 desiredAmount0,
+        uint256 desiredAmount1,
+        uint256 swapAmountIn,
+        uint256 swapMinAmountOut,
+        uint160 sqrtPriceLimitX96,
+        bool swapZeroForOne
+    ) internal {
+        _checkSwapLimits(sqrtPriceLimitX96, swapZeroForOne);
+
+        IERC20 t0Contract = IERC20(token0);
+        IERC20 t1Contract = IERC20(token1);
+
+        uint256 token0Balance = t0Contract.balanceOf(address(this));
+        uint256 token1Balance = t1Contract.balanceOf(address(this));
+
+        uint256 token0Needed = desiredAmount0 > token0Balance
+            ? desiredAmount0 - token0Balance
+            : 0;
+        uint256 token1Needed = desiredAmount1 > token1Balance
+            ? desiredAmount1 - token1Balance
+            : 0;
+
+        if (swapZeroForOne) {
+            // Amount available in reserve strategies
+            uint256 t1ReserveBal = IStrategy(poolTokens[token1].reserveStrategy)
+                .checkBalance(token1);
+
+            // Only swap when asset isn't available in reserve as well
+            require(
+                token1Needed > 0 && token1Needed > t1ReserveBal,
+                "Cannot swap when the asset is available in reserve"
+            );
+            // Additional amount of token0 required for swapping
+            token0Needed += swapAmountIn;
+            // Subtract token1 that we will get from swapping
+            token1Needed = (swapMinAmountOut >= token1Needed)
+                ? 0
+                : (token1Needed - swapMinAmountOut);
+
+            // Approve for swaps
+            t0Contract.safeApprove(address(swapRouter), swapAmountIn);
```

Comment:
(sorry commented on outdated code a few minutes earlier)...
should we call the safe approve only once instead on every swap?

Reply:
We would need to do it with max amount though...